python/ray/rllib/ddpg2/models.py

```diff
@@ -0,0 +1,408 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import numpy as np
+
+import tensorflow as tf
+import tensorflow.contrib.layers as layers
+
+from ray.rllib.models import ModelCatalog
+from ray.rllib.optimizers.multi_gpu_impl import TOWER_SCOPE_NAME
+
+
+def _build_s_network(registry, inputs, config):
+    frontend = ModelCatalog.get_model(registry, inputs, 1, config["model"])
+    return frontend.last_layer
+
+
+def _build_p_network(registry, inputs, dim_actions, config):
+    hiddens = config["actor_hiddens"]
+
+    action_out = inputs
+    for hidden in hiddens:
+        action_out = layers.fully_connected(
+            action_out, num_outputs=hidden, activation_fn=tf.nn.relu)
+    # The final output layer of the actor was a sigmoid layer, to bound the actions
+    # shape of action_scores is [batch_size, dim_actions]
+    action_scores = layers.fully_connected(
+        action_out, num_outputs=dim_actions, activation_fn=tf.nn.sigmoid)
+
+    return action_scores
+
+
+# As a stochastic policy for inference, but a deterministic policy for training
+# thus ignore batch_size issue when constructing a stochastic action
+def _build_action_network(
+        p_values, low_action, high_action, stochastic, eps, theta, sigma):
+    # shape is [None, dim_action]
+    deterministic_actions = (high_action-low_action) * p_values + low_action
+
+    exploration_sample = tf.get_variable(name="ornstein_uhlenbeck", dtype=tf.float32, \
```

Comment:
Every actor is only ever doing one rollout at a time, correct? Else you get  statefulness in the OU noise that assumes a single-process implementation

Reply:
yes, each time one observation only. There hasn't been any batch interaction functionality. For updating, policy is deterministic so the noise sample won't affect.